web/src/emoji_picker.ts

```diff
@@ -257,15 +274,54 @@ function filter_emojis() {
                 for (const alias of emoji_dict.aliases) {
                     const match = search_terms.every((search_term) => alias.includes(search_term));
                     if (match) {
-                        search_results.push({...emoji_dict, emoji_name: alias});
+                        const {reaction_type} = emoji.get_emoji_details_by_name(alias);
+                        // search_results array element is of type EmojiSuggestion which
+                        // is a union with `reaction_type` being the discriminator key. This means
+                        // we have to add objects separately depending on the type of discriminator key
+                        // to satisfy typescript. The assert statements are put to satisfy TS.
+
+                        if (emoji_dict.is_realm_emoji) {
+                            assert(reaction_type !== "unicode_emoji");
+                            search_results.push({
+                                ...emoji_dict,
+                                emoji_name: alias,
+                                reaction_type,
+                                type: "emoji",
+                            });
+                        } else {
+                            assert(reaction_type === "unicode_emoji");
+                            search_results.push({
+                                ...emoji_dict,
+                                emoji_name: alias,
+                                reaction_type,
+                                type: "emoji",
+                            });
+                        }
```

Comment:
This charade would be unnecessary with more accurate types.

```diff
diff --git a/web/shared/src/typeahead.ts b/web/shared/src/typeahead.ts
index b515c3b4dd..00f22fc5d7 100644
--- a/web/shared/src/typeahead.ts
+++ b/web/shared/src/typeahead.ts
@@ -314,7 +314,12 @@ export function triage<T>(
     };
 }
 
-export function sort_emojis<T extends EmojiSuggestion>(objs: T[], query: string): T[] {
+export type BaseEmoji = {emoji_name: string} & (
+    | {is_realm_emoji: true; emoji_code?: undefined}
+    | {is_realm_emoji: false; emoji_code: string}
+);
+
+export function sort_emojis<T extends BaseEmoji>(objs: T[], query: string): T[] {
     // replace spaces with underscores for emoji matching
     query = query.replace(/ /g, "_");
     query = query.toLowerCase();
@@ -326,9 +331,9 @@ export function sort_emojis<T extends EmojiSuggestion>(objs: T[], query: string)
 
     const popular_set = new Set(popular_emojis);
 
-    function is_popular(obj: EmojiSuggestion): boolean {
+    function is_popular(obj: BaseEmoji): boolean {
         return (
-            obj.reaction_type === "unicode_emoji" &&
+            !obj.is_realm_emoji &&
             popular_set.has(obj.emoji_code) &&
             decent_match(obj.emoji_name)
         );
@@ -364,7 +369,7 @@ export function sort_emojis<T extends EmojiSuggestion>(objs: T[], query: string)
     const unicode_emoji_codes = new Set();
     const sorted_unique_results: T[] = [];
     for (const emoji of sorted_results_with_possible_duplicates) {
-        if (emoji.reaction_type !== "unicode_emoji") {
+        if (emoji.is_realm_emoji) {
             sorted_unique_results.push(emoji);
         } else if (
             !unicode_emoji_codes.has(emoji.emoji_code) &&
--- a/web/src/emoji_picker.ts
+++ b/web/src/emoji_picker.ts
@@ -49,7 +48,7 @@ let emoji_catalog_last_coordinates = {
 let current_section = 0;
 let current_index = 0;
 let search_is_active = false;
-const search_results: EmojiSuggestion[] = [];
+const search_results: (EmojiDict & {emoji_name: string})[] = [];
 let section_head_offsets: {
     section: string;
     position_y: number;
```

Reply:
Added as a prep commit.

- https://github.com/zulip/zulip/pull/30459/commits/d0f75955bc608fe28a39d52ee51a88ff562de1b7